File: SC_cloud_assessment/SCcloud_pt2.py
# Como segunda etapa, gostaríamos de ver alguns códigos gerados por ti.
# Sendo assim, pedimos que tu crie códigos para os seguintes desafios.
# 1 - Fibonnaci
#     -- Criar uma função em sua linguagem preferida. A função deve receber um numero N >= 0 (deve validar o input para a função), e retornar o valor correspondente desse número na sequencia fibonnaci. EX. fib(0) =0; fib(1) = 1; fib(2) = 1; fib(3) = 2; fib(5) = 5; fib(6) = 8.
#     --- Criar uma função recursiva que resolva fibonacci
#     --- Criar uma função linear que resolva fibonnaci

def validate():
    while True:
        try:
            check_number = int(input("Qual n-ésimo dígito da sequência de Fibonacci deseja? "))
            if check_number > 0:
                # print(fibonacci_n) #test
                return check_number
            else:
                print("Inserir número positivo.")
        except ValueError:
            print("Inserir número positivo inteiro.")
     
def recursive_fibonacci(n):
    if n==1 or n==2:
        return 1
    else:
        return recursive_fibonacci(n-1) + recursive_fibonacci (n-2)

def linear_fibonacci(n):     
    if n==1 or n==2:
        return 1
    
    current = 1
    previous = 0
    for _ in range(2,n+1):
        new = current + previous
        previous = current
        current = new
    return current

valid_fibonacci = validate()
print("\nResolução da sequência de Fibonacci por função recursiva em Python.")
print(f'O valor do {valid_fibonacci}º número é {recursive_fibonacci(valid_fibonacci)}')

print("\nResolução da sequência de Fibonacci por função linear em Python.")
print(f'O valor do {valid_fibonacci}º número é {linear_fibonacci(valid_fibonacci)}\n')

File: SC_cloud_assessment/SCcloud_pt2.java
//  Como segunda etapa, gostaríamos de ver alguns códigos gerados por ti.
//  Sendo assim, pedimos que tu crie códigos para os seguintes desafios.
//  1 - Fibonnaci
//      -- Criar uma função em sua linguagem preferida. A função deve receber um numero N >= 0 (deve validar o input para a função), e retornar o valor correspondente desse número na sequencia fibonnaci. EX. fib(0) =0; fib(1) = 1; fib(2) = 1; fib(3) = 2; fib(5) = 5; fib(6) = 8.
//      --- Criar uma função recursiva que resolva fibonacci
//      --- Criar uma função linear que resolva fibonnaci

import java.util.Scanner;

public class Fibonacci {
    public static void main(String[] args) {
        int validFibonacci = validateInput();

        System.out.println("\nResolução da sequência de Fibonacci por função recursiva em Java.");
        System.out.println("O valor do " + validFibonacci + "º número é " + recursiveFibonacci(validFibonacci));

        System.out.println("\nResolução da sequência de Fibonacci por função linear em Java.");
        System.out.println("O valor do " + validFibonacci + "º número é " + linearFibonacci(validFibonacci));
    }

    public static int validateInput() {
        int checkNumber = 0;
        boolean isValidInput = false;
        Scanner scanner = new Scanner(System.in);

        while (!isValidInput) {
            try {
                System.out.print("Qual n-ésimo dígito da sequência de Fibonacci deseja? ");
                checkNumber = scanner.nextInt();

                if (checkNumber > 0) {
                    isValidInput = true;
                } else {
                    System.out.println("Inserir número positivo.");
                }
            } catch (java.util.InputMismatchException e) {
                System.out.println("Inserir número positivo inteiro.");
                scanner.nextLine();
            }
        }

        return checkNumber;
    }

    public static int recursiveFibonacci(int n) {
        if (n == 1 || n == 2) return 1;
        return recursiveFibonacci(n - 1) + recursiveFibonacci(n - 2);
    }

    public static int linearFibonacci(int n) {
        if (n == 1 || n == 2) return 1;

        int current = 1;
        int previous = 1;
        for (int i = 3; i <= n; i++) {
            int newFib = current + previous;
            previous = current;
            current = newFib;
        }

        return current;
    }
}
